### ISSUE TITLE
Create a custom vulkan memory allocator for Fuchsia's embedder

### ISSUE THREAD

**Issue by iskakaushik** [flutter/flutter#63024]
This is a blocker to get Flutter runner building in g3.

Currently skia relies on https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator to create a Vulkan memory allocator if we don't explicitly pass one to the background context. See: `flutter_runner::VulkanSurfaceProducer`.

Assuming we still want to use some `vk_mem_alloc` version, we would need to make a subclass of `GrVkMemoryAllocator.h` (note this is different from `GrVulkanMemoryAllocator.h`), create an object of that subclass and pass it into to Skia to use.

Basically the subclass would look very similar to what [GrVkAMDMemoryAllocator](https://fossies.org/linux/firefox/gfx/skia/skia/src/gpu/vk/GrVkAMDMemoryAllocator.cpp) is. With this approach we would have control over what version of `vk_mem_alloc` you are using and where we are getting it from as opposed to relying on skia's fork.

**Comment by iskakaushik** [comment]
I'm not exactly sure what the priority of this would be? Context, this lets us bisect failure for Fuchsia in Flutter rolls much more efficiently.

cc: @mehmetf , @kf6gpe

**Comment by kf6gpe** [comment]
I think this is something we should work on soon, to enable developer velocity in bug fixing. Assigning it a P3 for now.

**Comment by arbreng** [comment]
Will take a look at this soon after the Embedder API is complete

**Comment by chinmaygarde** [comment]
Building in g3 is not something the Flutter team provides support for. It seems like the issue here is that g3 and Skia disagree on the version of VMA that is suitable for use as the memory allocator. This disagreement needs to be resolved by the maintainers of Skia build rules in g3 and other consumers of Skia (and its dependencies) in g3. Creating and maintaining a custom memory allocator (and the various hooks for one in the engine) is not something we want in the Flutter engine.

Are there any other use cases for this beyond making it easier to bisect failures? For that, we want a model where failures are reported to the Flutter Team in the form of reduced test cases along with a reference to the version of Flutter that shows the failure. This is how errors are reported to Flutter from all other embedder implementations. Deploying engineering resources to bisect for errors in other codebases is not scalable.

**Comment by arbreng** [comment]
@chinmaygarde 

I think the intention of this work was to get fuchsia's embedder building in google3.  The custom memory allocator, if any, would live inside of the embedder's code.

We definitely don't want to deploy flutter team resources for this, or add any extra hooks to the engine if we can avoid it.  Perhaps this issue would be better raised on the Fuchsia bugtracker -- it's just a little sticky right now while ownership of the embedder is being transferred.

**Comment by mehmetf** [comment]
google3 cannot give reduced test cases on an issue that is blocking a roll, which is a P0. Building a reduced test case based on some complex application in google3 typically takes at least a day. g3 applications (and their dependencies) are very complex. The only quick action we can do is to find the culprit, revert it, then try to understand the failure before rolling it forward.

Plenty of Opal/Fuchsia developers would like the capability to test proposed engine changes in google3.

We also want to automate running engine tests in google3 for unmerged PRs. We are building automation tools around it. As binaries are not available on LUCI prior to a merge, the only way to test these would be to build the engine in google3.

**Comment by zanderso** [comment]
> We also want to automate running engine tests in google3 for unmerged PRs. We are building automation tools around it.

This is slightly off-topic, but I have concerns about the workflow for Flutter team members who are not Googlers in this situation. How will they debug tests that are only reported to fail in google3 on their PRs?

**Comment by mehmetf** [comment]
Every such PR (engine or framework) would need to have a Googler reviewer. That's needed in any case since the normal thing to do for PRs that break g3 is to revert them. At that point, a non-Googler would not be able to roll it forward. You would need someone who can see the failures in g3, understand what's going on and craft a proper test case.

**Comment by chinmaygarde** [comment]
@arbreng: Its fine to park this issue here till the bug tracking details are sorted out. I just wanted to make sure we didn't somehow sign up to write or maintain build rules in g3. The inability to select specific dependencies in g3 has meant that rules for other platforms in that environment have never been used to ship anything (with only Flutter vended pre-builts being used) and the existing rules (where available) have been maintained on a best-effort basis.

@mehmetf: Are there limitations to workflows where the engine is built using GN and that locally built engine being subbed into the g3 environment? This can be automated just fine. I am not sure why the engine has to built using blaze rules for the specific goals you listed to be accomplished. For native engine tests, if a test fails in a harness built using g3 rules but not using Flutter Engine build rules, the Flutter Team cannot provide support for it. This is because compilers, build flags, and dependencies might be different.

**Comment by chinmaygarde** [comment]
> For native engine tests, if a test fails in a harness built using g3 rules but not using Flutter Engine build rules, the Flutter Team cannot provide support for it. This is because compilers, build flags, and dependencies might be different.

The unstated suggestion here was that if you plan on adding automated tests that run harnesses not built using Flutter engine build rules, please don't. You are moving moving the responsibility of maintaining these rules onto whoever happens to be the sheriff managing the rolls at that time. This kind of scope creep is not something the team is equipped to handle. Of course, if a test in an engine built harness fails, that can be prioritized and fixed with utmost urgency.

**Comment by mehmetf** [comment]
My team has the responsibility of maintaining the build rules for engine in g3. We also have custom embedders that prefer to build the engine in g3 because they develop in g3. Having engine built in g3 *does* require changes to upstream code from time to time but these have always been small and good for code health (such as consistent imports).  If we end up in a scenario where engine code needs to be updated for the worse to support g3 rules, we can revisit this topic.

**Comment by mehmetf** [comment]
I also understand that compilers, build flags, and dependencies might be different and I agree it is a concern. That is why we are still importing prebuilts instead of switching everyone to g3 built binaries. However, from correctness standpoint, in the 1.5 years that we have been running automated tests on g3-built engine, I don't recall a single time where a failure was due to environmental differences. Dart VM and Skia move very quickly in g3 (just like they do on engine) and we keep the build flags fairly consistent.

**Comment by chinmaygarde** [comment]
> If we end up in a scenario where engine code needs to be updated for the worse to support g3 rules, we can revisit this topic.

The ask here is to add a custom memory allocator and its requisite hooks when there is no need for one within the engine. The only stated need so far is to support g3 rules. That's why we are visiting this topic now. David is also in the middle of the Embedder API migration. These hooks will also need to be exposed in that API for that migration to be completed. So this, if implemented, will become an ongoing maintenance burden to support.

> My team has the responsibility of maintaining the build rules for engine in g3.

That's great. Can we move this issue to this teams issue tracker? BTW, what are the roadblocks to using the version of VMA needed by Skia in g3? My sense is that this issue can be solved by correctly devising build rules and brining in the specific version of VMA needed by Skia as a third_party dependency. Since it seems like you are onboard with the idea of not making engine code worse to support these rules, I am leaning towards closing this issue here as wont-fix.

> ...  I don't recall a single time where a failure was due to environmental differences. 

Let's try to stay clear of hyperbole. I know I have had to figure out roll issues because of incorrect LTO flags were used by these variants for instance. So it is definitely > 0.

Can you also clarify my original query: Are there limitations to workflows where the engine is built using GN and that locally built engine being subbed into the g3 environment?

**Comment by mehmetf** [comment]
> That's why we are visiting this topic now. 

Ah. Thanks for clarifying. Yes, if this is a costly change that the engine can do without, pushing back is the right thing to do.

> Are there limitations to workflows where the engine is built using GN and that locally built engine being subbed into the g3 environment?

Infrastructure in google3 does not support GN. Flutter wants to support custom embedders in google3, which implies it needs to build with google3 infrastructure. For instance, customer:dream builds engine for a custom embedder for CastOS to run automated tests on TAP.

Dart and Fuchsia SDK developers also would like the ability to change code in google3 and run tests on Flutter. Asking all these engineers to build Flutter locally is a bigger collective burden than us maintaining engine build rules.

> My sense is that this issue can be solved by correctly devising build rules and brining in the specific version of VMA needed by Skia as a third_party dependency. Since it seems like you are onboard with the idea of not making engine code worse to support these rules, I am leaning towards closing this issue here as wont-fix.

Totally. If this is disruptive change, we should try to fix it in g3 first.

Before you mark as wont-fix, either @iskakaushik or @arbreng could perhaps explain why this proposal was opened in the first place rather than chasing a g3 specific fix.

**Comment by mehmetf** [comment]
@iskakaushik @arbreng 

While Skia g3 is being used in other iOS and Android apps, the only client of Skia g3 for Fuchsia is Flutter. Therefore, we have some leverage on customizing its build behavior and making sure the right set of dependencies are selected. Why would we not go that route?

**Comment by mehmetf** [comment]
@renyou will take a look at the flutter_runner issue in g3. This is good context for him and he will talk to @iskakaushik @arbreng separately.

**Comment by mehmetf** [comment]
Ren, if you can do: https://github.com/flutter/flutter/issues/63024#issuecomment-682832761, this issue can be closed.
